src/sentry/auth/staff.py

```diff
@@ -0,0 +1,319 @@
+        """
+        Returns ``(bool is_privileged, str reason)``
+        """
+        allowed_ips = self.allowed_ips
+
+        # _admin should have always completed SSO to gain status.
+        # We expect ORG_ID to always be set in production.
+        if ORG_ID and not has_completed_sso(self.request, ORG_ID):
+            if not DISABLE_SSO_CHECK_FOR_LOCAL_DEV:
+                return False, "incomplete-sso"
+
+        # if there's no IPs configured, we allow assume its the same as *
+        if not allowed_ips:
+            return True, None
+        ip = ipaddress.ip_address(str(self.request.META["REMOTE_ADDR"]))
+        if not any(ip in addr for addr in allowed_ips):
+            return False, "invalid-ip"
+        return True, None
+
+    def get_session_data(self, current_datetime=None):
+        """
+        Return the current session data, with native types coerced.
+        """
+        request = self.request
+        data = request.session.get(SESSION_KEY)
+
+        try:
+            cookie_token = request.get_signed_cookie(
+                key=COOKIE_NAME,
+                default=None,
+                salt=COOKIE_SALT,
+                max_age=MAX_STAFF_SESSION_AGE.total_seconds(),
+            )
+        except BadSignature:
+            logger.exception(
+                "staff.bad-cookie-signature",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+            )
+            return
+
+        if not cookie_token:
+            if data:
+                logger.warning(
+                    "staff.missing-cookie-token",
+                    extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+                )
+            return
+        elif not data:
+            logger.warning(
+                "staff.missing-session-data",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+            )
+            return
+
+        session_token = data.get("tok")
+        if not session_token:
+            logger.warning(
+                "staff.missing-session-token",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+            )
+            return
+
+        if not constant_time_compare(cookie_token, session_token):
+            logger.warning(
+                "staff.invalid-token",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+            )
+            return
+
+        if data["uid"] != str(request.user.id):
+            logger.warning(
+                "staff.invalid-uid",
+                extra={
+                    "ip_address": request.META["REMOTE_ADDR"],
+                    "user_id": request.user.id,
+                    "expected_user_id": data["uid"],
+                },
+            )
+            return
+
+        if current_datetime is None:
+            current_datetime = timezone.now()
```

Comment:
Is this safe? Should we be allowing for past or future checking?

Reply:
Hmm, honestly we can prob get rid of this in staff. I'm gonna leave it in for superuser in case something breaks, but for staff it should be fine